File: R/verify_pairing_counts.R
#!/usr/bin/env Rscript
# Base-R combinatorial check of the pairing table in
# proof/DEGREE3_TENSOR_IDENTITY.md.

pairings <- function(items) {
  if (length(items) == 0) return(list(list()))
  first <- items[[1]]
  output <- list()
  cursor <- 1
  for (index in 2:length(items)) {
    second <- items[[index]]
    keep <- setdiff(seq_along(items), c(1, index))
    tails <- pairings(items[keep])
    for (tail in tails) {
      output[[cursor]] <- c(list(list(first, second)), tail)
      cursor <- cursor + 1
    }
  }
  output
}

edge_key <- function(a, b) paste(sort(c(a, b)), collapse = "-")

classify_graph <- function(edges) {
  keys <- vapply(edges, function(edge) edge_key(edge[1], edge[2]), character(1))
  if (any(vapply(edges, function(edge) edge[1] == edge[2], logical(1)))) return("loop")
  counts <- table(keys)

  adjacency <- lapply(0:3, function(x) integer())
  for (edge in edges) {
    a <- edge[1] + 1
    b <- edge[2] + 1
    adjacency[[a]] <- unique(c(adjacency[[a]], b))
    adjacency[[b]] <- unique(c(adjacency[[b]], a))
  }
  seen <- 1L
  frontier <- 1L
  while (length(frontier) > 0) {
    current <- frontier[1]
    frontier <- frontier[-1]
    new_nodes <- setdiff(adjacency[[current]], seen)
    seen <- unique(c(seen, new_nodes))
    frontier <- c(frontier, new_nodes)
  }
  if (length(seen) < 4) return("D")
  if (length(counts) == 6 && all(as.integer(counts) == 1)) return("K")
  "J"
}

count_case <- function(free_degrees, fixed_edges) {
  items <- list()
  cursor <- 1
  for (vertex in 0:3) {
    degree <- free_degrees[vertex + 1]
    if (degree > 0) {
      for (slot in 0:(degree - 1)) {
        items[[cursor]] <- c(vertex, slot)
        cursor <- cursor + 1
      }
    }
  }
  counts <- c(loop = 0L, D = 0L, J = 0L, K = 0L)
  for (matching in pairings(items)) {
    moment_edges <- lapply(matching, function(pair) c(pair[[1]][1], pair[[2]][1]))
    type <- classify_graph(c(fixed_edges, moment_edges))
    counts[type] <- counts[type] + 1L
  }
  counts
}

odd_double_factorial <- function(n) {
  stopifnot(n > 0, n %% 2 == 1)
  prod(seq(1, n, by = 2))
}

cases <- list(
  "Y^4" = list(c(3, 3, 3, 3), list()),
  "Y^2|v|^2" = list(c(3, 3, 2, 2), list(c(2, 3))),
  "Y^2|M|^2" = list(c(3, 3, 1, 1), list(c(2, 3), c(2, 3))),
  "|v|^4" = list(c(2, 2, 2, 2), list(c(0, 1), c(2, 3))),
  "|v|^2|M|^2" = list(c(2, 2, 1, 1), list(c(0, 1), c(2, 3), c(2, 3))),
  "|M|^4" = list(c(1, 1, 1, 1), list(c(0, 1), c(0, 1), c(2, 3), c(2, 3)))
)

expected <- list(
  "Y^4" = c(D = 108, J = 1944, K = 1296, tau = 756, Jred = 648),
  "Y^2|v|^2" = c(D = 12, J = 216, K = 144, tau = 84, Jred = 72),
  "Y^2|M|^2" = c(D = 6, J = 36, K = 0, tau = 6, Jred = 36),
  "|v|^4" = c(D = 4, J = 40, K = 16, tau = 12, Jred = 24),
  "|v|^2|M|^2" = c(D = 2, J = 8, K = 0, tau = 2, Jred = 8),
  "|M|^4" = c(D = 1, J = 2, K = 0, tau = 1, Jred = 2)
)

cat("R PAIRING-COUNT CHECK\n")
for (name in names(cases)) {
  values <- cases[[name]]
  counts <- count_case(values[[1]], values[[2]])
  stopifnot(sum(counts) == odd_double_factorial(sum(values[[1]]) - 1))
  n_tau <- counts["D"] + counts["K"] / 2
  j_reduced <- counts["J"] - counts["K"]
  actual <- c(D = counts["D"], J = counts["J"], K = counts["K"], tau = n_tau, Jred = j_reduced)
  stopifnot(all(actual == expected[[name]]))
  cat(sprintf("%-17s D=%4d J=%4d K=%4d => tau^2=%4d, J=%4d\n",
              name, as.integer(counts["D"]), as.integer(counts["J"]),
              as.integer(counts["K"]), as.integer(n_tau), as.integer(j_reduced)))
}
cat("R PAIRING-COUNT CHECK: PASS\n")

File: python/verify_pairing_counts.py
#!/usr/bin/env python3
"""Exact combinatorial verification of the quartic pairing-count table.

The script enumerates all pairings of the free sphere-moment half-edges for the
six quartic integrands in proof/DEGREE3_TENSOR_IDENTITY.md.  Surviving complete
contractions are classified as D (tau^2), J (tr(B^2)), or K (tetrahedral).
The dimension-three identity K=tau^2/2-J is then applied.
"""

from __future__ import annotations

from collections import Counter
from fractions import Fraction
from math import prod
from typing import Iterable, Iterator

HalfEdge = tuple[int, int]
Edge = tuple[int, int]


def pairings(items: tuple[HalfEdge, ...]) -> Iterator[tuple[tuple[HalfEdge, HalfEdge], ...]]:
    """Generate all perfect pairings of a tuple of labeled half-edges."""
    if not items:
        yield ()
        return
    first = items[0]
    for index in range(1, len(items)):
        second = items[index]
        rest = items[1:index] + items[index + 1 :]
        for tail in pairings(rest):
            yield ((first, second),) + tail


def classify(edges: Iterable[Edge]) -> str:
    """Classify a complete cubic contraction graph on four tensor vertices."""
    counts = Counter(tuple(sorted(edge)) for edge in edges)
    if any(a == b for a, b in counts):
        return "loop"

    adjacency = {vertex: set() for vertex in range(4)}
    for a, b in counts:
        adjacency[a].add(b)
        adjacency[b].add(a)

    seen = {0}
    stack = [0]
    while stack:
        vertex = stack.pop()
        for neighbor in adjacency[vertex]:
            if neighbor not in seen:
                seen.add(neighbor)
                stack.append(neighbor)

    if len(seen) < 4:
        return "D"
    if len(counts) == 6 and all(multiplicity == 1 for multiplicity in counts.values()):
        return "K"
    return "J"


def odd_double_factorial(number: int) -> int:
    if number < 1 or number % 2 == 0:
        raise ValueError("Expected a positive odd integer.")
    return prod(range(1, number + 1, 2))


def count_case(free_degrees: tuple[int, int, int, int], fixed_edges: tuple[Edge, ...]) -> Counter[str]:
    half_edges: list[HalfEdge] = []
    for vertex, degree in enumerate(free_degrees):
        half_edges.extend((vertex, slot) for slot in range(degree))

    result: Counter[str] = Counter()
    for matching in pairings(tuple(half_edges)):
        moment_edges = [(left[0], right[0]) for left, right in matching]
        result[classify((*fixed_edges, *moment_edges))] += 1
    return result


def main() -> None:
    cases = {
        "Y^4": ((3, 3, 3, 3), ()),
        "Y^2|v|^2": ((3, 3, 2, 2), ((2, 3),)),
        "Y^2|M|^2": ((3, 3, 1, 1), ((2, 3), (2, 3))),
        "|v|^4": ((2, 2, 2, 2), ((0, 1), (2, 3))),
        "|v|^2|M|^2": ((2, 2, 1, 1), ((0, 1), (2, 3), (2, 3))),
        "|M|^4": ((1, 1, 1, 1), ((0, 1), (0, 1), (2, 3), (2, 3))),
    }

    expected = {
        "Y^4": (108, 1944, 1296, 756, 648),
        "Y^2|v|^2": (12, 216, 144, 84, 72),
        "Y^2|M|^2": (6, 36, 0, 6, 36),
        "|v|^4": (4, 40, 16, 12, 24),
        "|v|^2|M|^2": (2, 8, 0, 2, 8),
        "|M|^4": (1, 2, 0, 1, 2),
    }

    reduced: dict[str, tuple[int, int, int]] = {}
    print("PAIRING-COUNT CHECK")
    for name, (degrees, fixed) in cases.items():
        counts = count_case(degrees, fixed)
        free_slots = sum(degrees)
        total_expected = odd_double_factorial(free_slots - 1)
        if sum(counts.values()) != total_expected:
            raise AssertionError(f"Pairing total failed for {name}.")

        n_d = counts["D"]
        n_j = counts["J"]
        n_k = counts["K"]
        n_tau = Fraction(n_d) + Fraction(n_k, 2)
        n_j_reduced = n_j - n_k
        if n_tau.denominator != 1:
            raise AssertionError(f"Nonintegral reduced tau count for {name}.")

        actual = (n_d, n_j, n_k, int(n_tau), n_j_reduced)
        if actual != expected[name]:
            raise AssertionError(f"Unexpected counts for {name}: {actual}")
        reduced[name] = (free_slots // 2, int(n_tau), n_j_reduced)
        print(
            f"{name:17s} D={n_d:4d} J={n_j:4d} K={n_k:4d} "
            f"=> tau^2={int(n_tau):4d}, J={n_j_reduced:4d}"
        )

    coefficients: dict[str, tuple[Fraction, Fraction]] = {}
    for name, (m, n_tau, n_j_reduced) in reduced.items():
        denominator = odd_double_factorial(2 * m + 1)
        coefficients[name] = (
            Fraction(4 * n_tau, denominator),
            Fraction(4 * n_j_reduced, denominator),
        )

    expected_coefficients = {
        "Y^4": (Fraction(16, 715), Fraction(96, 5005)),
        "Y^2|v|^2": (Fraction(16, 495), Fraction(32, 1155)),
        "Y^2|M|^2": (Fraction(8, 315), Fraction(16, 105)),
        "|v|^4": (Fraction(16, 315), Fraction(32, 315)),
        "|v|^2|M|^2": (Fraction(8, 105), Fraction(32, 105)),
        "|M|^4": (Fraction(4, 15), Fraction(8, 15)),
    }
    if coefficients != expected_coefficients:
        raise AssertionError("Moment coefficient table failed.")

    weights = {
        "Y^4": 4096,
        "Y^2|v|^2": 18432,
        "Y^2|M|^2": -9216,
        "|v|^4": 20736,
        "|v|^2|M|^2": -20736,
        "|M|^4": 5184,
    }
    tau_coefficient = sum(Fraction(weights[name]) * coefficients[name][0] for name in cases)
    j_coefficient = sum(Fraction(weights[name]) * coefficients[name][1] for name in cases)
    if tau_coefficient != Fraction(1_310_464, 1001):
        raise AssertionError(f"Unexpected tau^2 coefficient: {tau_coefficient}")
    if j_coefficient != Fraction(-2_265_600, 1001):
        raise AssertionError(f"Unexpected J coefficient: {j_coefficient}")

    b0_tau_coefficient = tau_coefficient + j_coefficient / 3
    if b0_tau_coefficient != Fraction(555_264, 1001):
        raise AssertionError(f"Unexpected B0-form tau coefficient: {b0_tau_coefficient}")

    print(f"q^2 / pi coefficient of tau^2 = {tau_coefficient}")
    print(f"q^2 / pi coefficient of J     = {j_coefficient}")
    print(f"B0-form coefficient of tau^2  = {b0_tau_coefficient}")
    print("PAIRING-COUNT CHECK: PASS")


if __name__ == "__main__":
    main()
